flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/source/FileStoreSourceSplitReader.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.store.connector.source;
+
+import org.apache.flink.connector.base.source.reader.RecordsWithSplitIds;
+import org.apache.flink.connector.base.source.reader.splitreader.SplitReader;
+import org.apache.flink.connector.base.source.reader.splitreader.SplitsAddition;
+import org.apache.flink.connector.base.source.reader.splitreader.SplitsChange;
+import org.apache.flink.connector.file.src.impl.FileRecords;
+import org.apache.flink.connector.file.src.reader.BulkFormat;
+import org.apache.flink.connector.file.src.util.MutableRecordAndPosition;
+import org.apache.flink.connector.file.src.util.Pool;
+import org.apache.flink.connector.file.src.util.RecordAndPosition;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.store.file.KeyValue;
+import org.apache.flink.table.store.file.operation.FileStoreRead;
+import org.apache.flink.table.store.file.utils.RecordReader;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.Queue;
+
+/** The {@link SplitReader} implementation for the file store source. */
+public class FileStoreSourceSplitReader
+        implements SplitReader<RecordAndPosition<RowData>, FileStoreSourceSplit> {
+
+    private final FileStoreRead fileStoreRead;
+    private final boolean keyAsRecord;
+
+    private final Queue<FileStoreSourceSplit> splits;
+
+    private final Pool<FileStoreRecordIterator> pool;
+
+    @Nullable private RecordReader currentReader;
+    @Nullable private String currentSplitId;
+    private long currentNumRead;
+    private RecordReader.RecordIterator currentFirstBatch;
+
+    public FileStoreSourceSplitReader(FileStoreRead fileStoreRead, boolean keyAsRecord) {
+        this.fileStoreRead = fileStoreRead;
+        this.keyAsRecord = keyAsRecord;
+        this.splits = new LinkedList<>();
+        this.pool = new Pool<>(1);
+        this.pool.add(new FileStoreRecordIterator());
+    }
+
+    @Override
+    public RecordsWithSplitIds<RecordAndPosition<RowData>> fetch() throws IOException {
+        checkSplitOrStartNext();
+
+        // pool first, avoid thread safety issues
```

Comment:
Not thread safety. This is because batches can only be fetched one by one, so we use a pool to restrict the reader thread not to fetch too many batches at the same time.

Reply:
`pool first, pool size is 1, the underlying implementation does not allow multiple batches to be read at the same time`